### ISSUE TITLE
Allow modification of batch for `on_batch_start` hook

### ISSUE THREAD

**Issue by kevinlu1211** [PyTorchLightning/pytorch-lightning#1733]
## 🚀 Feature
Allow modification of batch for `on_batch_start` hook in the `LightningModule`

### Motivation
For the `LightningModule` there is an `on_batch_start` method that is called. Right now it either returns `-1` signifying that training should not continue, or `None`. There are scenarios whereby I want to change the batch that is being returned by the `DataLoader` before it is being transferred to the GPU.

### Pitch

Let the `on_batch_start` hook on the `LightningModule` return two variables:
1. The batch
2. Whether or not to continue training

### Alternatives

The workaround that I currently have is:

```
def on_batch_start(self, batch):
    batch.to_pytorch_lightning_format()
```

Where the `to_pytorch_lightning_format()` will alter the `Batch` object such that it mimics a dictionary so that the PTL will transfer the Tensors onto the GPU. Prior to doing this, since PTL only accepts recognizes lists, tuples, and dicts the Tensors in my `Batch` object will not be transferred onto the GPU.

Though this isn't very desirable as it causes a side effect, ideally what I want to do is something like:

```
def on_batch_start(self, batch):
    batch = batch.to_pytorch_lightning_format()
    return batch
```

There are other use cases that I've encountered:
1. More fine-grained control over which sample in the batch is to be used

I've read through the relevant source code, and it seems like a simple fix (if I'm not mistaken). I'd be happy to make a PR if this is something that is aligned with the development of PTL.

**Comment by awaelchli** [comment]
I'm not sure on_batch_start is the right place to provide this functionality. Looks like you want to do kind of the same as in #1245. There I proposed to add a transfer hook in the LightningModule. What do you think of that? Would it be appropriate for your use case?

**Comment by williamFalcon** [comment]
this sounds like you are looking for transforms... anything about formatting a batch should be done in the transforms going into the dataset

**Comment by kevinlu1211** [comment]
Not sure what you mean by transforms, but I have overridden the default `collate_fn` for the Dataloader to return a Batch object. This object is used by the callbacks I have defined to do various things e.g. recording the training samples that have a high loss during an epoch etc. I used to use `fastai` in the past which allowed the modification of the batch object during the `on_batch_begin` method. 

@awaelchli yes it would be appropriate though it still does seem like a hacky fix, as there are other scenarios whereby the batch may need to be modified before training. Would it not make more sense to give the user full control in the `on_batch_start` hook like `fastai` does?

**Comment by dscarmo** [comment]
Best practice in PyTorch is to have your Dataset class do all the transforms needed over the data.

After the data leaves your dataset, in my opnion, you shouldn't try to change it. The DataLoader should only handle collating the samples into a batch and parallelism (num_workers). 

Its very inefficient to modify data in the main loop.
